goalie/optimisation.py

```diff
@@ -151,19 +169,73 @@ def minimise(self):
+                    if ee_converged:
+                        pyrint("Error estimator convergence detected.")
+                        if mesh_seq.params.convergence_criteria == "any":
+                            mesh_seq.converged[:] = True
+                            self.adaptive = False
+                else:
+                    ee_converged = True
+
+                # Adapt meshes and log element counts
+                continue_unconditionally = self.adaptor(
+                    mesh_seq,
+                    mesh_seq.solutions,
+                    mesh_seq.indicators,
+                    **self.adaptor_kwargs,
+                )
+                if mesh_seq.params.drop_out_converged:
+                    mesh_seq.check_convergence[:] = np.logical_not(
+                        np.logical_or(continue_unconditionally, mesh_seq.converged)
+                    )
+                mesh_seq.element_counts.append(mesh_seq.count_elements())
+                mesh_seq.vertex_counts.append(mesh_seq.count_vertices())
+
+                # Check for element count convergence
+                mesh_seq.converged[:] = mesh_seq.check_element_count_convergence()
+                elem_converged = mesh_seq.converged.all()
+                if elem_converged:
+                    pyrint("Element count convergence detected.")
+                    if mesh_seq.params.convergence_criteria == "any":
+                        self.adaptive = False
+
+                # Convergence check for 'all' mode
+                if qoi_converged and ee_converged and elem_converged:
+                    pyrint("Convergence of all quantities detected.")
+                    self.adaptive = False
+
             # Update initial condition getter for the next iteration
-            ics = self.mesh_seq.get_initial_condition()
-            ics[self.control] = u
-            if self.mesh_seq._get_initial_condition is None:
+            ics = mesh_seq.get_initial_condition()
+            ics[self.control] = Function(
+                mesh_seq.function_spaces[self.control][0], val=float(u)
+            )
+            if mesh_seq._get_initial_condition is None:
                 # NOTE:
                 # * mesh_seq.get_initial_condition may have been defined directly in the
                 #   'object-oriented' approach (for example, see
                 #   https://mesh-adaptation.github.io/docs/demos/burgers_oo.py.html)
                 # * Ruff raises 'B023 Function definition does not bind loop variable
                 #   `ics`' but this is intentional so we suppress the error.
-                self.mesh_seq.get_initial_condition = lambda: ics  # noqa: B023
+                mesh_seq.get_initial_condition = lambda: ics  # noqa: B023
             else:
-                self.mesh_seq._get_initial_condition = lambda mesh_seq: ics  # noqa: B023
+                mesh_seq._get_initial_condition = lambda mesh_seq: ics  # noqa: B023
 
             # Check for convergence and divergence
```

Comment:
We should format the gradient convergence check the same way as the mesh adaptation convergence checks.

Reply:
Done in e60353f149a721cd440ad79ae63cf99f1ea1926e.